src/actors/workers/blocks.rs

```diff
@@ -85,27 +91,57 @@ where
         Ok(blocks)
     }
 
+    /// Collect blocks according to the predicate `fun` and send those blocks to
+    ///  the metadata actor.
+    async fn collect_and_send(&self, fun: impl Fn(u32) -> bool + Send + 'static) -> Result<()> {
+        self.meta
+            .send(BatchBlock::new(self.collect_blocks(fun).await?))
+            .await?;
+        Ok(())
+    }
+
     /// First run of indexing
     /// gets any blocks that are missing from database and indexes those.
     /// sets the `last_max` value.
-    async fn re_index(&mut self) -> Result<Option<Vec<Block<B>>>> {
+    async fn re_index(&mut self) -> Result<()> {
         let mut conn = self.db.send(GetState::Conn.into()).await?.await?.conn();
-        let numbers = queries::missing_blocks_min_max(&mut conn, self.last_max).await?;
+        let mut numbers = queries::missing_blocks_min_max(&mut conn, self.last_max).await?;
         let len = numbers.len();
         log::info!("{} missing blocks", len);
         self.last_max = if let Some(m) = queries::max_block(&mut conn).await? {
             m
         } else {
             // a `None` means that the blocks table is not populated yet
-            return Ok(None);
+            return Ok(());
         };
-        let blocks = self.collect_blocks(move |n| numbers.contains(&n)).await?;
-        Ok(Some(blocks))
+
+        if numbers.len() as u32 > self.max_block_load {
+            let mut tmp_set = HashSet::<u32>::with_capacity(self.max_block_load as usize);
```

Comment:
I wonder if we could switch to returning a `BTreeSet` from `missing_blocks_min_max` and use `split_off` on the result? Or even limit `missing_blocks_min_max` to `max_block_load` to fetch the batch of blocks we need?